system-helper/flatpak-system-helper.c

```diff
@@ -1618,36 +1544,36 @@ check_partial_pull_availability (const gchar   *repo_tmp,
           break;
         }
 
-      if (file_info == NULL)
+      if (file_info == NULL || res == TRUE)
         break;
 
       name = g_file_info_get_name (file_info);
       if (g_file_info_get_file_type (file_info) == G_FILE_TYPE_DIRECTORY &&
-          g_str_has_prefix (name, prefix))
+          g_str_has_prefix (name, "flatpak-cache-"))
         {
-          g_autoptr(GFile) cache_dir = g_file_get_child (repo_tmpfile, name);
-          gboolean dir_in_use = FALSE;
+          g_autoptr(GFile) cache_dir_file = g_file_get_child (repo_tmpfile, name);
+          g_autofree gchar *cache_dir_name = g_file_get_path (cache_dir_file);
 
-          /* If this cache directory is used by any other ongoing pull, we
-           * should not be re(using) it. */
-          G_LOCK (ongoing_pulls);
-          for (guint i = 0; i < ongoing_pulls->len; i++)
+          G_LOCK (cache_dirs_in_use);
+          if (!g_hash_table_lookup_extended (cache_dirs_in_use, cache_dir_name, NULL, NULL))
             {
-              OngoingPull *pull = g_ptr_array_index (ongoing_pulls, i);
-              if (g_strcmp0 (pull->src_dir, name) == 0)
+              struct stat st_buf;
+
+              /* We are able to find a cache dir which is not in use. */
+              if (stat (cache_dir_name, &st_buf) == 0 &&
+                  st_buf.st_uid == passwd->pw_uid &&        /* should be owned by SYSTEM_HELPER_USER */
+                  (st_buf.st_mode & 0022) == 0)             /* should not be world-writeable */
                 {
-                  dir_in_use = TRUE;
-                  break;
+                  g_hash_table_insert (cache_dirs_in_use, g_strdup (cache_dir_name), NULL);
```

Comment:
Perhaps:

```c
gboolean did_not_exist = g_hash_table_insert (cache_dirs_in_use, g_strdup (cache_dir_name), NULL);
g_assert (did_not_exist);
```

Reply:
Review accepted and addressed in the latest round of iteration.